// react-device-frameset/lib/DeviceOptions.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceNames = exports.DeviceOptions = exports.defineDevice = void 0;
var defineDevice = function (definition) { return definition; };
exports.defineDevice = defineDevice;
exports.DeviceOptions = (_a = {},
    _a['iPhone X'] = (0, exports.defineDevice)({
        device: 'iphone-x',
        colors: [],
        hasLandscape: true,
        width: 375,
        height: 812,
    }),
    _a['iPhone 8'] = (0, exports.defineDevice)({
        device: 'iphone8',
        colors: ['black', 'silver', 'gold'],
        hasLandscape: true,
        width: 375,
        height: 667,
    }),
    _a['iPhone 8 Plus'] = (0, exports.defineDevice)({
        device: 'iphone8plus',
        colors: ['black', 'silver', 'gold'],
        hasLandscape: true,
        width: 414,
        height: 736,
    }),
    _a['iPhone 5s'] = (0, exports.defineDevice)({
        device: 'iphone5s',
        colors: ['black', 'silver', 'gold'],
        hasLandscape: true,
        width: 320,
        height: 568,
    }),
    _a['iPhone 5c'] = (0, exports.defineDevice)({
        device: 'iphone5c',
        colors: ['white', 'red', 'yellow', 'green', 'blue'],
        hasLandscape: true,
        width: 320,
        height: 568,
    }),
    _a['iPhone 4s'] = (0, exports.defineDevice)({
        device: 'iphone4s',
        colors: ['black', 'silver'],
        hasLandscape: true,
        width: 320,
        height: 480,
    }),
    _a['Galaxy Note 8'] = (0, exports.defineDevice)({
        device: 'note8',
        colors: [],
        hasLandscape: true,
        width: 400,
        height: 822,
    }),
    _a['Nexus 5'] = (0, exports.defineDevice)({
        device: 'nexus5',
        colors: [],
        hasLandscape: true,
        width: 320,
        height: 568,
    }),
    _a['Lumia 920'] = (0, exports.defineDevice)({
        device: 'lumia920',
        colors: ['black', 'white', 'yellow', 'red', 'blue'],
        hasLandscape: true,
        width: 320,
        height: 533,
    }),
    _a['Samsung Galaxy S5'] = (0, exports.defineDevice)({
        device: 's5',
        colors: ['white', 'black'],
        hasLandscape: true,
        width: 320,
        height: 568,
    }),
    _a['HTC One'] = (0, exports.defineDevice)({
        device: 'nexus5',
        colors: [],
        hasLandscape: true,
        width: 320,
        height: 568,
    }),
    _a['iPad Mini'] = (0, exports.defineDevice)({
        device: 'ipad',
        colors: ['black', 'silver'],
        hasLandscape: true,
        width: 576,
        height: 768,
    }),
    _a['MacBook Pro'] = (0, exports.defineDevice)({
        device: 'macbook',
        colors: [],
        hasLandscape: false,
        width: 960,
        height: 600,
    }),
    _a);
exports.DeviceNames = Object.keys(exports.DeviceOptions);
