// react-device-frameset/lib/DeviceEmulator.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceEmulator = void 0;
var react_1 = __importStar(require("react"));
var DeviceFrameset_1 = require("./DeviceFrameset");
var DeviceOptions_1 = require("./DeviceOptions");
var zooms = [
    0.5, 0.75, 1, 1.25, 1.5
];
exports.DeviceEmulator = react_1.default.memo(function DeviceEmulator(_a) {
    var _b;
    var children = _a.children, value = _a.value, onChange = _a.onChange, _c = _a.banDevices, banDevices = _c === void 0 ? [] : _c, divProps = __rest(_a, ["children", "value", "onChange", "banDevices"]);
    var deviceNames = (0, react_1.useMemo)(function () { return DeviceOptions_1.DeviceNames.filter(function (devName) { return !banDevices.includes(devName); }); }, []);
    var _d = (0, react_1.useState)((_b = deviceNames[0]) !== null && _b !== void 0 ? _b : ''), deviceName = _d[0], setDeviceName = _d[1];
    var selectedDeviceName = (0, react_1.useMemo)(function () { var _a; return (_a = value === null || value === void 0 ? void 0 : value.device) !== null && _a !== void 0 ? _a : deviceName; }, [value, deviceName]);
    var handleSelectChange = (0, react_1.useCallback)(function (event) {
        var newDeviceName = event.target.value;
        if (!deviceNames.includes(newDeviceName))
            throw new Error("Invalid device name for " + newDeviceName);
        setDeviceName(newDeviceName);
    }, [deviceNames]);
    var _e = (0, react_1.useState)(zooms[2]), selectedZoom = _e[0], setSelectedZoom = _e[1];
    var handleSelectZoomChange = (0, react_1.useCallback)(function (event) {
        var newZoom = +event.target.value;
        if (!zooms.includes(newZoom))
            throw new Error("Invalid device zoom for " + newZoom);
        setSelectedZoom(newZoom);
    }, [deviceNames]);
    var _f = (0, react_1.useMemo)(function () { return DeviceFrameset_1.DeviceOptions[selectedDeviceName]; }, [selectedDeviceName]), colors = _f.colors, hasLandscape = _f.hasLandscape, width = _f.width, height = _f.height;
    var firstColor = (0, react_1.useMemo)(function () { return colors[0]; }, [colors]);
    var _g = (0, react_1.useState)(undefined), isLandscape = _g[0], setIsLandscape = _g[1];
    var isLandscapeChecked = (0, react_1.useMemo)(function () { return hasLandscape ? isLandscape : undefined; }, [hasLandscape, isLandscape]);
    var handleIsLandscapeChange = (0, react_1.useCallback)(function () {
        if (!hasLandscape)
            return;
        setIsLandscape(function (is) { return !is; });
    }, [hasLandscape]);
    var deviceFramesetProps = (0, react_1.useMemo)(function () { return value !== null && value !== void 0 ? value : ({
        device: selectedDeviceName,
        color: firstColor,
        landscape: isLandscapeChecked,
        width: width,
        height: height,
        zoom: selectedZoom,
    }); }, [value, selectedDeviceName, firstColor, isLandscapeChecked, width, height, selectedZoom]);
    (0, react_1.useEffect)(function () {
        onChange === null || onChange === void 0 ? void 0 : onChange(deviceFramesetProps);
    }, [onChange, deviceFramesetProps]);
    return (react_1.default.createElement("div", __assign({ className: "device-emulator" }, divProps),
        react_1.default.createElement("section", null,
            react_1.default.createElement("select", { value: selectedDeviceName, onChange: handleSelectChange }, deviceNames.map(function (devName) { return (react_1.default.createElement("option", { key: devName, value: devName }, devName)); })),
            react_1.default.createElement("input", { disabled: true, value: width }),
            react_1.default.createElement("span", null, "x"),
            react_1.default.createElement("input", { disabled: true, value: height }),
            react_1.default.createElement("select", { value: selectedZoom, onChange: handleSelectZoomChange }, zooms.map(function (zoom) { return (react_1.default.createElement("option", { key: zoom, value: zoom },
                zoom * 100,
                "%")); })),
            react_1.default.createElement("label", null, "Landscape:"),
            react_1.default.createElement("input", { type: "checkbox", checked: !!isLandscapeChecked, disabled: !hasLandscape, onChange: handleIsLandscapeChange })),
        react_1.default.createElement("div", { className: "device-emulator-container" }, children(deviceFramesetProps))));
});
